lib/http_aws_sigv4.c

```diff
@@ -154,6 +154,66 @@ static int compare_header_names(const char *a, const char *b)
   return cmp;
 }
 
+/* merge duplicate header definitions by comma delimiting their values
+   in the order defined the headers are defined */
+static CURLcode merge_duplicate_headers(struct curl_slist *head)
+{
+  struct curl_slist *curr = head;
+
+  while(curr) {
+    struct curl_slist *next = curr->next;
+    if(!next) {
+      break;
+    }
+
+    if(compare_header_names(curr->data, next->data) == 0) {
+      char *colon_next;
+      char *merged;
+      char *val_next;
+      size_t curr_len;
+      size_t val_next_len;
+      size_t new_len;
+
+      colon_next = strchr(next->data, ':');
+
+      DEBUGASSERT(colon_next);
+
+      curr_len = strlen(curr->data);
+      val_next = colon_next + 1;
+
+      val_next_len = strlen(val_next);
+
+      /* add 1 for the comma we'll insert */
+      new_len = curr_len + 1 + val_next_len;
+
+      merged = malloc(new_len + 1); /* add 1 for null terminator */
```

Comment:
I would prefer if this was written to use a `struct dynbuf` instead to build the new string in, to avoid the malloc and direct memcpy calls.

Reply:
Updated to use dynbuf